### ISSUE TITLE
Messaging.messaging().token returns errors

### ISSUE THREAD

**Issue by RobinCaroff** [firebase/firebase-ios-sdk#10679]
### Description

It has been two days since that last Messaging.messaging().token success with our app. 

It currently returns the following error :

```
Error Domain=com.google.fcm Code=0 "Invalid fetch response, expected 'token' or 'Error' key" UserInfo={NSLocalizedFailureReason=Invalid fetch response, expected 'token' or 'Error' key}
```

Here is how we retrieve the token :

```
self.getMessaging().token { token, error in
              if let error = error {
                Logger.shared.log(.app, .error, "Error fetching FCM registration token: \(error)")
                observer(.failure(error))
              } else if let token = token {
                Logger.shared.log(.app, .debug, "FCM registration token: \(token)")
                observer(.success(token))
              }
            }
            return Disposables.create()
```

Note 1: the live app has not been updated since 2 weeks and we reproduce the same issue with oldest version.

Note 2: We have a different target using the same code with a different bundle id which does not have the issue. 

### Firebase SDK Version

10.3.0

### Xcode Version

14.1

### Installation Method

CocoaPods

### Firebase Product(s)

Messaging

### Targeted Platforms

iOS

### Relevant Log Output

```shell
Error Domain=com.google.fcm Code=0 "Invalid fetch response, expected 'token' or 'Error' key" UserInfo={NSLocalizedFailureReason=Invalid fetch response, expected 'token' or 'Error' key}
```

### If using Swift Package Manager, the project's Package.resolved

<details>
<summary>Expand <code>Package.resolved</code> snippet</summary>
<br>

```json

Replace this line with the contents of your Package.resolved.

```

</details>

### If using CocoaPods, the project's Podfile.lock

<details>
<summary>Expand <code>Podfile.lock</code> snippet</summary>
<br>

```yml

PODS:
  - Alamofire (5.6.4)
  - Differentiator (5.0.0)
  - Firebase/AnalyticsWithoutAdIdSupport (10.3.0):
    - Firebase/CoreOnly
    - FirebaseAnalytics/WithoutAdIdSupport (~> 10.3.0)
  - Firebase/CoreOnly (10.3.0):
    - FirebaseCore (= 10.3.0)
  - Firebase/Crashlytics (10.3.0):
    - Firebase/CoreOnly
    - FirebaseCrashlytics (~> 10.3.0)
  - Firebase/Messaging (10.3.0):
    - Firebase/CoreOnly
    - FirebaseMessaging (~> 10.3.0)
  - FirebaseAnalytics/WithoutAdIdSupport (10.3.0):
    - FirebaseCore (~> 10.0)
    - FirebaseInstallations (~> 10.0)
    - GoogleAppMeasurement/WithoutAdIdSupport (= 10.3.0)
    - GoogleUtilities/AppDelegateSwizzler (~> 7.8)
    - GoogleUtilities/MethodSwizzler (~> 7.8)
    - GoogleUtilities/Network (~> 7.8)
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
    - nanopb (< 2.30910.0, >= 2.30908.0)
  - FirebaseCore (10.3.0):
    - FirebaseCoreInternal (~> 10.0)
    - GoogleUtilities/Environment (~> 7.8)
    - GoogleUtilities/Logger (~> 7.8)
  - FirebaseCoreInternal (10.3.0):
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
  - FirebaseCrashlytics (10.3.0):
    - FirebaseCore (~> 10.0)
    - FirebaseInstallations (~> 10.0)
    - GoogleDataTransport (~> 9.2)
    - GoogleUtilities/Environment (~> 7.8)
    - nanopb (< 2.30910.0, >= 2.30908.0)
    - PromisesObjC (~> 2.1)
  - FirebaseInstallations (10.3.0):
    - FirebaseCore (~> 10.0)
    - GoogleUtilities/Environment (~> 7.8)
    - GoogleUtilities/UserDefaults (~> 7.8)
    - PromisesObjC (~> 2.1)
  - FirebaseMessaging (10.3.0):
    - FirebaseCore (~> 10.0)
    - FirebaseInstallations (~> 10.0)
    - GoogleDataTransport (~> 9.2)
    - GoogleUtilities/AppDelegateSwizzler (~> 7.8)
    - GoogleUtilities/Environment (~> 7.8)
    - GoogleUtilities/Reachability (~> 7.8)
    - GoogleUtilities/UserDefaults (~> 7.8)
    - nanopb (< 2.30910.0, >= 2.30908.0)
  - GoogleAppMeasurement/WithoutAdIdSupport (10.3.0):
    - GoogleUtilities/AppDelegateSwizzler (~> 7.8)
    - GoogleUtilities/MethodSwizzler (~> 7.8)
    - GoogleUtilities/Network (~> 7.8)
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
    - nanopb (< 2.30910.0, >= 2.30908.0)
  - GoogleDataTransport (9.2.0):
    - GoogleUtilities/Environment (~> 7.7)
    - nanopb (< 2.30910.0, >= 2.30908.0)
    - PromisesObjC (< 3.0, >= 1.2)
  - GoogleUtilities/AppDelegateSwizzler (7.11.0):
    - GoogleUtilities/Environment
    - GoogleUtilities/Logger
    - GoogleUtilities/Network
  - GoogleUtilities/Environment (7.11.0):
    - PromisesObjC (< 3.0, >= 1.2)
  - GoogleUtilities/Logger (7.11.0):
    - GoogleUtilities/Environment
  - GoogleUtilities/MethodSwizzler (7.11.0):
    - GoogleUtilities/Logger
  - GoogleUtilities/Network (7.11.0):
    - GoogleUtilities/Logger
    - "GoogleUtilities/NSData+zlib"
    - GoogleUtilities/Reachability
  - "GoogleUtilities/NSData+zlib (7.11.0)"
  - GoogleUtilities/Reachability (7.11.0):
    - GoogleUtilities/Logger
  - GoogleUtilities/UserDefaults (7.11.0):
    - GoogleUtilities/Logger
  - IQKeyboardManager (6.5.10)
  - Kingfisher (7.4.1)
  - Moya (15.0.0):
    - Moya/Core (= 15.0.0)
  - Moya/Core (15.0.0):
    - Alamofire (~> 5.0)
  - Moya/RxSwift (15.0.0):
    - Moya/Core
    - RxSwift (~> 6.0)
  - nanopb (2.30909.0):
    - nanopb/decode (= 2.30909.0)
    - nanopb/encode (= 2.30909.0)
  - nanopb/decode (2.30909.0)
  - nanopb/encode (2.30909.0)
  - NSLogger/ObjC (1.9.7)
  - NSLogger/Swift (1.9.7):
    - NSLogger/ObjC
  - PromisesObjC (2.1.1)
  - RxBlocking (6.5.0):
    - RxSwift (= 6.5.0)
  - RxCocoa (6.5.0):
    - RxRelay (= 6.5.0)
    - RxSwift (= 6.5.0)
  - RxDataSources (5.0.0):
    - Differentiator (~> 5.0)
    - RxCocoa (~> 6.0)
    - RxSwift (~> 6.0)
  - RxGesture (4.0.4):
    - RxCocoa (~> 6.0)
    - RxSwift (~> 6.0)
  - RxRelay (6.5.0):
    - RxSwift (= 6.5.0)
  - RxSwift (6.5.0)
  - RxSwiftExt (6.0.1):
    - RxSwiftExt/Core (= 6.0.1)
    - RxSwiftExt/RxCocoa (= 6.0.1)
  - RxSwiftExt/Core (6.0.1):
    - RxSwift (~> 6.0)
  - RxSwiftExt/RxCocoa (6.0.1):
    - RxCocoa (~> 6.0)
    - RxSwiftExt/Core
  - RxTest (6.5.0):
    - RxSwift (= 6.5.0)
  - SwiftLint (0.49.1)
  - SwiftyAttributes (5.1.1)

DEPENDENCIES:
  - Firebase/AnalyticsWithoutAdIdSupport
  - Firebase/Crashlytics
  - Firebase/Messaging
  - IQKeyboardManager (~> 6.5.9)
  - Kingfisher (~> 7.4.1)
  - Moya (~> 15.0.0)
  - Moya/RxSwift (~> 15.0.0)
  - NSLogger/Swift
  - RxBlocking (~> 6.5.0)
  - RxCocoa (~> 6.5.0)
  - RxDataSources (~> 5.0.0)
  - RxGesture (~> 4.0.4)
  - RxSwift (~> 6.5.0)
  - RxSwiftExt (~> 6.0.1)
  - RxTest (~> 6.5.0)
  - SwiftLint (~> 0.49.1)
  - SwiftyAttributes (~> 5.1.1)

SPEC REPOS:
  https://github.com/CocoaPods/Specs.git:
    - Alamofire
    - Differentiator
    - Firebase
    - FirebaseAnalytics
    - FirebaseCore
    - FirebaseCoreInternal
    - FirebaseCrashlytics
    - FirebaseInstallations
    - FirebaseMessaging
    - GoogleAppMeasurement
    - GoogleDataTransport
    - GoogleUtilities
    - IQKeyboardManager
    - Kingfisher
    - Moya
    - nanopb
    - NSLogger
    - PromisesObjC
    - RxBlocking
    - RxCocoa
    - RxDataSources
    - RxGesture
    - RxRelay
    - RxSwift
    - RxSwiftExt
    - RxTest
    - SwiftLint
    - SwiftyAttributes

SPEC CHECKSUMS:
  Alamofire: 4e95d97098eacb88856099c4fc79b526a299e48c
  Differentiator: e8497ceab83c1b10ca233716d547b9af21b9344d
  Firebase: f92fc551ead69c94168d36c2b26188263860acd9
  FirebaseAnalytics: 036232b6a1e2918e5f67572417be1173576245f3
  FirebaseCore: 988754646ab3bd4bdcb740f1bfe26b9f6c0d5f2a
  FirebaseCoreInternal: 29b76f784d607df8b2a1259d73c3f04f1210137b
  FirebaseCrashlytics: f20d956f8229010b645e534693c39e0b7843c268
  FirebaseInstallations: e2f26126089dcf41e215f7b8925af8d953c7d602
  FirebaseMessaging: e345b219fd15d325f0cf2fef28cb8ce00d851b3f
  GoogleAppMeasurement: c7d6fff39bf2d829587d74088d582e32d75133c3
  GoogleDataTransport: 1c8145da7117bd68bbbed00cf304edb6a24de00f
  GoogleUtilities: c2bdc4cf2ce786c4d2e6b3bcfd599a25ca78f06f
  IQKeyboardManager: 45a1fa55c1a5b02c61ac0fd7fd5b62bb4ad20d97
  Kingfisher: cd762a593a61b2fbecf7645c00f9a801a3ebfc9c
  Moya: 138f0573e53411fb3dc17016add0b748dfbd78ee
  nanopb: b552cce312b6c8484180ef47159bc0f65a1f0431
  NSLogger: cf3013a4fba189f631f07e6c3d5b8ce2b209654b
  PromisesObjC: ab77feca74fa2823e7af4249b8326368e61014cb
  RxBlocking: 04b5fd28bb5ea49f7d64b80db8bbfe50d9cc1c7d
  RxCocoa: 94f817b71c07517321eb4f9ad299112ca8af743b
  RxDataSources: aa47cc1ed6c500fa0dfecac5c979b723542d79cf
  RxGesture: f3efb47ed2d26a8082f7b660d4a59970e275a7f8
  RxRelay: 1de1523e604c72b6c68feadedd1af3b1b4d0ecbd
  RxSwift: 5710a9e6b17f3c3d6e40d6e559b9fa1e813b2ef8
  RxSwiftExt: d61c4b9b06e0a448041b1c659e0f1b5c5a035afc
  RxTest: eb2d23adefc5a5ebf5779c7792fa3edfe6ebcc17
  SwiftLint: 32ee33ded0636d0905ef6911b2b67bbaeeedafa5
  SwiftyAttributes: e3e13891ddac3c62b3d7798d5c3969cea88e62bb

PODFILE CHECKSUM: e2e890938470dccb8e731dbb945147be1c8c971c

COCOAPODS: 1.11.3


```

</details>

**Comment by RobinCaroff** [comment]
Similar issue: https://github.com/invertase/react-native-firebase/issues/6830

**Comment by tomwanzek** [comment]
@RobinCaroff Thanks for opening this one and cross-referencing 🙇 

While not directly related, it seems contextually interesting that we have started seeing some other previously non-existent behaviors with the Firebase Messaging SDK/React Native Firebase Messaging.

We were doing some regression testing for an upgrade of our RN Instabug integration in the App, and encountered another errors on iOS, which we had never seen "surfaced" to our error reporting:

* `[messaging/unknown] The Internet connection appears to be offline.` This message was logged as part of a scenario were we deliberately took down the network (router offline for Wifi). Historically, I suspect, this might have been handled "silently" by the Messaging SDK?

Again, just for context if that helps narrow down how the `messaging.getToken()` issue made it into a codebase that had not updated Firebase dependencies when the error started to occur.

**Comment by henryadebayo** [comment]
having the same issue as @RobinCaroff  how do i resolve it ?

**Comment by RobinCaroff** [comment]
> Hi all, I tried to reproduce the issue but I was able to retrieve the token successfully. With this, could you share a minimal repro app as well as the complete stack trace or app debug logs for further investigation?

Hi @rizafran . Thank you for your answer. What's really strange is that I also use the same codebase with a different bundle id (my preprod environment) and I don't get the error. I checked the Firebase plist but it is up to date.

I'll try to produce a minimal repo where I can reproduce the error. In the meantime, is there any specific Firebase log that I could provide to help you narrow down the issue?

**Comment by boseidel** [comment]
Also ran into this. I have tried various versions of the firebase pod during the installation of the plugin and still facing the error on emulators as well as production environments.

**Comment by flaviodsilverio** [comment]
> What's really strange is that I also use the same codebase with a different bundle id (my preprod environment) and I don't get the error. I checked the Firebase plist but it is up to date.

This makes it even more curious, since the same exact thing is happening to me. In one of the environments the works flawlessly, the other is failing constantly!

**Comment by fiizzy** [comment]
Having the same issue. No idea what's happening. Started all of a sudden.

**Comment by XavierMF** [comment]
Same issue here. We notice an error spike on firebase crashlytics as we send the error
<img width="621" alt="Capture d’écran 2023-01-17 à 22 25 56" src="https://user-images.githubusercontent.com/604382/213019898-4eff4e96-1b93-4759-952c-bd270ed51e77.png">

**Comment by fiizzy** [comment]
Can we say this is a firebase thingy for sure?

**Comment by Seanmclem** [comment]
Maybe Certs are expired because of the new year? Maybe that's just me.
I went to my Project, in the Firebase Console, Project Settings, Cloud Messaging, scrolled down to where the Apple APN keys are setup, and my Dev and Prod ones are invalid. One in December, and one Today. If I replace those, this will probably start working again

**Comment by henryadebayo** [comment]
@Seanmclem chaged APNs key Ali but it did not fix the issue maybe it may work you

**Comment by Seanmclem** [comment]
@henryadebayo Just tried it too, and no change either.

**Comment by yeahse** [comment]
So serious problem

**Comment by Seanmclem** [comment]
@henryadebayo I was able to fix it based on your reply, so thanks! However, I was thinking about the plist thing -and felt like it was unlikely that _all_ of our `GoogleServices-info.plist` files are suddenly corrupt in the past 5 days. So rather than deleting it -I did the other part you mentioned. I stopped and restarted my app. Not even a full rebuild, just a new JS bundle (I'm using React Native), so a quick stop and restart.

The only other thing we have in common is we both just re-added our APN certificates. So that might be required too, but afterwards restarting the build was definitely the final thing that fixed it for me.

**Comment by redccoma** [comment]
We are using Firebase Unity SDK 9.40, which wraps Firebase iOS SDK, and we are experiencing the same problem as the issue in this article, so we leave an opinion.

In the past week, 2 out of 15 apps in service were only on iOS.
  - All apps use the same code
  - App crash occurs while the app starts with the error described in this article

The method we implemented was implemented in items 1 and 2 below.

1. Obtain a token by registering an event like the guide at https://firebase.google.com/docs/cloud-messaging/unity/client#initialize

2. If the token was not obtained in method 1, acquire the token by additionally implementing the https://stackoverflow.com/a/69222193/13368553 method

In one app, the app crash was fixed by removing method 2, but in the other app, even if method 2 was removed, it was not fixed, so the code implemented in #1 was deleted to respond.

Today, we're facing the same issue with the rest of the apps, and we're going to respond the same way we finally responded.

However, if you respond in this way, you will not be able to use FCM, so you need to respond quickly.

**Comment by yeahse** [comment]
> @Seanmclem and @yeahse so i removed the GoogleServices-info.plist file from my Runner folder, then i did flutter clean and clean build folder then flutter pub get and it worked, yet to replace the GoogleServices-info.plist file , i i'm thinking the GoogleServices-info.plist file is corrupt don't know how or if this ever happens will replace it and let you know if it still works

I just followed your solution and it does not work

**Comment by XavierMF** [comment]
An update has been made to the Firebase SDK to provide an APN token at launch on the latest SDK release (https://firebase.google.com/support/release-notes/ios)
Obviously, this modification applies even if we are not using the latest version of the SDK. This also applies to older versions which should not be the case....

**Comment by Monsteel** [comment]
@XavierMF How can I follow this method? I saw the [document](https://firebase.google.com/docs/cloud-messaging/ios/client#access_the_registration_token), but I don't know.

**Comment by XavierMF** [comment]
If you have an APN token (user has accepted push notification), it must be forwarded to firebase sdk. And then you should be able to retrieve a firebase token.
However, if you use this token not only for enable push notification user, you won't be able to retrieve a token and that is the problem. If you use older version of the SDK, this rules apply also... It shouldn't be the cause and that's why this is a Firebase issue from my point of view

**Comment by rodkiewicz** [comment]
The same problem occurs with flutter applications using [firebase_messaging](https://pub.dev/packages/firebase_messaging)

**Comment by Tahir-7cTech** [comment]
facing same issue in my flutter app. FCM just stopped working with the following error
`
FLTFirebaseMessaging: An error occurred while calling method Messaging#getToken, errorOrNil => {
    NSLocalizedFailureReason = "Invalid fetch response, expected 'token' or 'Error' key";
}
[VERBOSE-2:dart_vm_initializer.cc(41)] Unhandled Exception: [firebase_messaging/unknown] An unknown error has occurred.
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:653:7)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:315:18)
<asynchronous suspension>
#2      MethodChannel.invokeMapMethod (package:flutter/src/services/platform_channel.dart:518:43)
<asynchronous suspension>
#3      MethodChannelFirebaseMessaging.getToken (package:firebase_messaging_platform_interface/src/method_channel/method_channel_messaging.dart:224:11)
<asynchronous suspension>
#4      _MyAppState.requestAndRegisterNotification (package:myapp/main.dart:174:23)
`

**Comment by rizafran** [comment]
Hi all, I'll be raising this to the team, and see what we can do here. Thanks.

**Comment by ProgrammerFaraz** [comment]
Unable to get fcm token in some iOS Devices.

**Comment by fiizzy** [comment]
I need to confirm this, 

> Does anyone have the error on a physical device in production ?

cause it seem to me that it the error has to do with Simulator not being a physical device and returns `null` value as its APN token🤔

**Comment by boseidel** [comment]
> I need to confirm this,
> 
> > Does anyone have the error on a physical device in production ?
> 
> cause it seem to me that the error has to do with Simulator not being a physical device and returns `null` value as its APN token🤔

Yes, this is happening in production on IOS devices. Both simulators and real devices are causing the same error.

**Comment by Monsteel** [comment]
I solved it❗️

In fact, the notification `deviceID(APNs Key)` injection time was slower than the time to call `token(_:)`.

Like you, it happened a few days ago.

I want you to check the timing of the initialization.

**Comment by codewithdo** [comment]
I am having this error as well! What's the issue?

**Comment by henryadebayo** [comment]
@Monsteel please tell how you solved it, and are you using react or flutter ?

**Comment by Seanmclem** [comment]
> > I need to confirm this,
> 
> > 
> 
> > > Does anyone have the error on a physical device in production ?
> 
> > 
> 
> > cause it seem to me that the error has to do with Simulator not being a physical device and returns `null` value as its APN token🤔
> 
> 
> 
> Yes, this is happening in production on IOS devices. Both simulators and real devices are causing the same error.

Simulator can get an FCM token, just not real notifications. Although with this exact error in this exact scenario, I solved it by re-adding new APN Certs in firebase messaging console, and restarting my device with a new build.

**Comment by AymericLeFeyer** [comment]
I have the same issue in my flutter app. Before it was working well.
`firebase_messaging: ^14.0.3`
```
FLTFirebaseMessaging: An error occurred while calling method Messaging#subscribeToTopic, errorOrNil => {
    NSLocalizedFailureReason = "Invalid fetch response, expected 'token' or 'Error' key";
}
```

**Comment by adirgan** [comment]
So what is the solution, do you have to get the APN token first, and then try to get the FCM token?

**Comment by adirgan** [comment]
@aashishpatil-g It is understood, but they cannot make a change so drastic that it breaks applications that are already in production without sending a warning that they will make a change that can break functionality if they do not do the following

**Comment by vitalii-tym** [comment]
The [change log for Version 10.4.0 - January 17, 2023](https://firebase.google.com/support/release-notes/ios) says:

> On app startup, an APNs Token must be provided to FCM SDK before retrieving an FCM Token, otherwise an error will be returned as part of the completion.

I tried this and it seems to be working for me as of Firebase v8.15:
<img width="621" alt="Screenshot 2023-01-18 at 19 58 37" src="https://user-images.githubusercontent.com/5410835/213258509-e82e8cf6-b6dd-42cd-a4d2-9cfc0a1d6c05.png">

Not sure if this is the right thing to do though. I hope Firebase updates their documentation to match the recent changes.

P.S. I also tried to find and upload my old APNs Auth Key and downloaded the `GoogleServices-info.plist` file (which turned out to be exactly same), therefore I don't know if any of these actions were required in addition to the change in code I mentioned above.

**Comment by jfbourne** [comment]
1+ following

**Comment by MuriloSantosLima** [comment]
> 

swizzling shouldn't send apns token to fcm ?

**Comment by suyuti** [comment]
Firebase  Apple SDK released 17th Jan 2023.
FCM initialization has been changed.
APNs Token must be provided before getToken otherwise returns error.

[release notes](https://firebase.google.com/support/release-notes/ios#version_1040_-_january_17_2023)

But I could not get APNSToken thouh. Returns null always on both real device and simulator. What about you? How do you get APNsToken?

**Comment by suyuti** [comment]
> If you have an APN token (user has accepted push notification), it must be forwarded to firebase sdk. And then you should be able to retrieve a firebase token. However, if you use this token not only for enable push notification user, you won't be able to retrieve a token and that is the problem. If you use older version of the SDK, this rules apply also... It shouldn't be the cause and that's why this is a Firebase issue from my point of view

getAPNsToken() always returns null on both real device and simulators. How did you get?

**Comment by Monsteel** [comment]
> @Monsteel please tell how you solved it, and are you using react or flutter ? 

Using Swift. My app is Native app.

`UNUserNotificationCenterDelegate` began to receive late.

**Comment by Monsteel** [comment]
> @Monsteel are you using Swift?
> 
> 

Yes I using Swift.

**Comment by mikehardy** [comment]
Hey hopefully not too terrible form to direct tag @paulb777 but 🙏  help 🙏  - the react-native-firebase test app reproduces this as well, in our messaging e2e API exercise, on iOS simulator, but only on our deleteToken API test (!?). At the same time we have lots of users hollering at us (and have been for 5 days) but we don't have guidance on how exactly to

a) make sure we have a valid APNS token set in messaging and/or if not
b) how to get an APNS token if needed, prior to calling deleteToken or getToken as needed

**Comment by aashishpatil-g** [comment]
This server change has been pulled back for now. It might take a few hours for it to propagate.

**Comment by DanielReyesDev** [comment]
@aashishpatil-g , our issue (https://github.com/firebase/firebase-ios-sdk/issues/10710) has gone after rollback.

**Comment by adirgan** [comment]
@aashishpatil-g What they have to do is that only SDK 10.4.0 and above is the one that does not allow obtaining the FCM token without the APN, so that this change has to be implemented from that version, because applying it to everyone is not ethical

In my case the problem is already solved, but I will update to 10.4.0 with the new way they ask to obtain the FCM Token

**Comment by aashishpatil-g** [comment]
@DanielReyesDev - thanks for confirming that the issue was resolved. 

@adirgan  - thanks for the feedback.

**Comment by codewithdo** [comment]
It worked, thanks!

**Comment by raulnoz** [comment]
Solved here

**Comment by XavierMF** [comment]
Up here also. @aashishpatil-g  if you can give us more infos about the case of this evolution will be live again. Which SDK version, when, etc...
However thank you for reactivity

**Comment by sveyp** [comment]
Solved as well! Thanks!

**Comment by RobinCaroff** [comment]
It works also on my app. Thanks !

**Comment by Tahir-7cTech** [comment]
Thanks issue is fixed

**Comment by suyuti** [comment]
It works as expected now.

**Comment by aashishpatil-g** [comment]
Following up with some guidance for the Firebase 10.4.0 SDK. 

As stated in the [release notes](https://firebase.google.com/support/release-notes/ios#fcm), 10.4.0 onwards, we are declining vending a FCM token until an APNS token is provided. We have always been logging a WARNING when this has occurred in previous versions of the SDK but haven’t declined to provide a FCM token. Such a vended FCM token is not useful since it is missing a target APNS token and thus has no endpoint to send the pushes to. 

Scenarios: 

1. On a fresh install of the app, 
Firebase configure flow triggers the FCM register flow even though there isn’t a APNS token. You will see an error message due to this flow. This error can be ignored as the FCM token is refetched once an APNS token is provided. I am investigating if this can be fixed in a future SDK release. 

2. For Objective C or Swift (not SwiftUI) projects, if swizzling is enabled, 
FCM token refetch happens automatically when we get the APNS token. The swizzling will take care of listening for when APNS token is available. The FCM messaging delegate callback will be called once an FCM token is available. 

func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?)

3. If swizzling is disabled or if your project is pure SwiftUI (swizzling does not work well with SwiftUI projects) 
You must implement the UIApplicationDelegate callbacks to listen for APNS token availability. Check the SwiftUI sample in the FCM SDK codebase for how to achieve this. 

4. Testing: If you have unit or integration tests running in the simulator, it is not possible to get an APNS token unless specific environment is used - iOS 16 Simulator + macOS 13 + Apple Silicon HW. If that environment condition is not met, there isn’t an APNS token available in iOS simulator. As a result, any tests for fetching FCM token will error out in the Firebase 10.4.0 SDK. Either reconsider the tests or try to provide fake APNS token for test purposes only. If you provide a fake APNS token, be sure to delete any FCM token created from this process. The FIRMessagingPubSubTest in the Firebase SDK uses a fake APNS token.

**Comment by kishan-thinkbridge** [comment]
I have solved this issue by mentioning a specific version of the firebase core in **Podfile** and the rest of the firebase pods will adopt the respective version which will convenient to them.

`pod 'Firebase/Core', '8.12.1'`

Remove these files from the project:
- Podfile.lock
- Pods [Directory]
- .xcworkspace

and just go with the "pod install". 
That's it.

**Comment by kishanbarmawala** [comment]
I have solved this issue by mentioning a specific version of the firebase core in **Podfile** and the rest of the firebase pods will adopt the respective version which will convenient to them.

`pod 'Firebase/Core', '8.12.1'`

Remove these files from the project:
- Podfile.lock
- Pods [Directory]
- .xcworkspace

and just go with the "pod install". 
That's it.

**Comment by mikehardy** [comment]
But... @kishanbarmawala this reads the same as "I have decided to continue creating invalid FCM tokens in our project, by permanently relying on outdated versions of the software" - I mean that in a technical sense, I believe those are technically valid substitutions in your statement. I totally understand if it is just deferring the maintenance to a future date when project timelines allow but I would be careful adopting that version pin permanently...

**Comment by aashishpatil-g** [comment]
The issue identified in bullet 1 in the [comment](https://github.com/firebase/firebase-ios-sdk/issues/10679#issuecomment-1402776795) has been fixed in 10.6.0 - https://firebase.google.com/support/release-notes/ios#fcm

I will close this issue since there aren't any opens.
